peps/pep-9999.rst

```diff
@@ -0,0 +1,856 @@
+      - List of identifiers this entry connects to.
+        Useful to annotate aliases or virtual package implementations.
+        If no `provides` info is added, the entry is considered canonical.
+      - False
+    * - ``urls``
+      - ``AnyUrl | list[AnyUrl] | dict[NonEmptyString, AnyUrl]``
+      - Hyperlinks to web locations that provide more information about the definition.
+      - False
+
+Known ecosystems schema
+^^^^^^^^^^^^^^^^^^^^^^^
+
+The known ecosystems list is specified by the following
+`JSON Schema <https://github.com/jaimergp/external-metadata-mappings/blob/main/schemas/known-ecosystems.schema.json>`__:
+
+``$schema``
+~~~~~~~~~~~
+
+.. list-table::
+    :widths: 25 75
+
+    * - Title
+      - $Schema
+    * - Type
+      - ``string``
+    * - Description
+      - URL of the mappings schema in use for the document.
+    * - Required
+      - False
+
+``schema_version``
+~~~~~~~~~~~~~~~~~~
+
+.. list-table::
+    :widths: 25 75
+
+    * - Title
+      - Schema Version
+    * - Type
+      - ``integer``
+    * - Required
+      - False
+
+``ecosystems``
+~~~~~~~~~~~~~~
+
+.. list-table::
+    :widths: 25 75
+
+    * - Title
+      - Ecosystems
+    * - Type
+      - ``dict``
+    * - Description
+      - Ecosystems names and their corresponding details.
+    * - Required
+      - True
+
+This dictionary maps non-empty string keys referring to the ecosystem identifiers
+to a sub-dictionary defined as:
+
+.. list-table::
+    :header-rows: 1
+    :widths: 20 25 40 15
+
+    * - Field
+      - URL
+      - Description
+      - Required
+    * - ``schema``
+      - ``AnyURL``
+      - URL to the mapping for this ecosystem
+      - True
+    * - ``package_managers``
+      - ``list[string]``
+      - List of package manager names known in this ecosystem
+      - True
+
+Mappings schema
+^^^^^^^^^^^^^^^
+
+The mappings are specified by the following
+`JSON Schema <https://github.com/jaimergp/external-metadata-mappings/blob/main/schemas/external-mapping.schema.json>`__:
+
+``$schema``
+~~~~~~~~~~~
+
+.. list-table::
+    :widths: 25 75
+
+    * - Title
+      - $Schema
+    * - Type
+      - ``string``
+    * - Description
+      - URL of the mappings schema in use for the document.
+    * - Required
+      - False
+
+``schema_version``
+~~~~~~~~~~~~~~~~~~
+
+.. list-table::
+    :widths: 25 75
+
+    * - Title
+      - Schema Version
+    * - Type
+      - ``integer``
+    * - Required
+      - False
+
+``name``
+~~~~~~~~
+
+.. list-table::
+    :widths: 25 75
+
+    * - Title
+      - Name
+    * - Type
+      - ``string``
+    * - Description
+      - Name of the schema
+    * - Required
+      - True
+
+``description``
+~~~~~~~~~~~~~~~
+
+.. list-table::
+    :widths: 25 75
+
+    * - Title
+      - Description
+    * - Type
+      - ``string | None``
+    * - Description
+      - Free-form field to add information this mapping. Allows
+        Markdown.
+    * - Required
+      - False
+
+``mappings``
```

Comment:
One open question is whether to make `mappings` a:

- `list[dict]` for each mapping entry, where each entry specifies the `id` (PURL) it refers to
- `dict[str, list[dict]` keyed by `id`, where the values are the different mapping entries associated to that PURL.

Reply:
Right now it's using `list[dict]` out of simplicity, but it requires iterating the index to find entries. Not that access performance matters at the current scale, but it might in the future. Also the Python API could group by ID internally for performance, if needed.

Reply:
I'm not sure without looking at the code, but it feels like we may have O(10) lookups for O(100,000) length package lists, so it may matter and it'd be best to be sure there won't be a performance issue at scale.

Reply:
For those cases, couldn't the consuming tool just regroup the mappings in a more convenient way? What I'm mostly after here is a convenient way for folks to annotate and generate the mappings.

Shall I generate arbitrarily large mappings to see how the different strategies play out?

Reply:
Might be useful. Unless you prefer the `dict[str, list[dict]` anyway of course.